scripts/generate-lib.js

```diff
@@ -0,0 +1,322 @@
+    ++i;
+  }
+
+  return Promise.all(promises).then(() => {
+    log(write, "");
+    console.timeEnd(logLabel);
+  });
+};
+
+/**
+ * Generate lib files for the client
+ */
+const libClient = async () => {
+  log(console.log, "CLIENT: Starting client lib generation");
+  console.time("Finished client lib generation");
+
+  let steps = [];
+
+  // Set NODE_ENV manually if not set as babel-preset-react-app depends on it
+  if (!process.env.NODE_ENV) {
+    log(
+      console.log,
+      "CLIENT: INFO: NODE_ENV was empty, setting NODE_ENV=production"
+    );
+    process.env.NODE_ENV = "production";
+  }
+
+  // (1) remove the lib directories, if they exist
+  if (fs.existsSync(clientLibPath)) {
+    await asyncRimraf(clientLibPath);
+  }
+  await fs.promises.mkdir(clientLibPath);
+
+  // (2) copy non-source files, so imports don't error out
+  steps.push(
+    new Promise((res) => {
+      glob(
+        "**/*.!(ts|tsx|svg)",
+        { cwd: clientSrcPath, nodir: true },
+        async (err, files) => {
+          if (err) throw err;
+
+          // ignore test files
+          files = files.filter(
+            (p) => !p.includes("__tests__") && !p.includes("__mocks__")
+          );
+
+          await makeDirs(files, clientLibPath);
+
+          const promises = new Array(files.length);
+          const logLabel = `Copied ${files.length} non-source files`;
+          let i = 0;
+
+          console.time(logLabel);
+          for (const file of files) {
+            promises[i] = new Promise((res) => {
+              fs.copyFile(
+                path.resolve(clientSrcPath, file),
+                path.resolve(clientLibPath, file),
+                (err) => {
+                  if (err) throw err;
+                  res();
+                }
+              );
+            });
+            ++i;
+          }
+
+          Promise.all(promises).then(() => {
+            log(write, "CLIENT: ");
+            console.timeEnd(logLabel);
+            res();
+          });
+        }
+      );
+    })
+  );
+
+  // (3) run babel on the typescript source files
+  babel.loadPartialConfig({
+    configFile: path.resolve(__dirname, "..", ".babelrc.json"),
+  });
+
+  steps.push(
+    new Promise((res) => {
+      glob("**/*.{ts,tsx}", { cwd: clientSrcPath }, (err, files) => {
+        if (err) throw err;
+
+        // ignore test files
+        files = files.filter(
+          (p) => !p.includes("__tests__") && !p.includes("__mocks__")
+        );
+
+        transpileSources(
+          clientSrcPath,
+          clientLibPath,
+          files,
+          "CLIENT: Compiled {} source files",
+          (code, file) => {
+            // Replace calls to ESM exports with calls to CommonJS exports
+            code = code.replace(
+              /@babel(\\\\|\/)runtime(\\\\|\/)helpers(\\\\|\/)esm/g,
+              "@babel$1runtime$2helpers"
+            );
+
+            // Replace API root to point to backend
+            if (file === "api/api.ts") {
```

Comment:
I think  we can have the actual `api.ts` use the environment variable. If we use something like `REACT_APP_SERVER_URL`, it'll get frozen in at build time.

Reply:
I tried that, but the issue is that the variable substitution shows up in the built files that we serve to the client. Even worse, the CRA webpack settings actually bundle the value of `process` at build-time, so we'd have to edit those settings, which we can't.

I can make the substitution more robust though. 30e5a44